### ISSUE TITLE
Bug: [no-unnecessary-condition] false positive on optional private field

### ISSUE THREAD

**Issue by JacobLey** [typescript-eslint/typescript-eslint#9564]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.3&fileType=.tsx&code=MYGwhgzhAEDCD2A7CBLCAXApo9AlMiAJtAN4BQ00AxAE4GED8AXNIgK4C2ARpjQNxkK0AOaZ0sMMAAWmQviIAKAJSkhldFLQA6WvWgMGAXmgBZMBq10i8DsoGVKNMWxqJoG7bqL3oAXzL%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2ToAPAIbNcBYkNQAFTLERY8ATwAqi%2BWUod0vAL4gdQA&tokens=false

### Repro Code

```TypeScript
class ConsistentRand {
  #rand?: number;

  getCachedRand() {
    this.#rand ??= Math.random();
    return this.#rand;
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true,
    "exactOptionalPropertyTypes": true
  }
}
```

### Expected Result

The example code should not be an error for `no-unnecessary-condition`.

The private field is marked optional and typescript acknowledges that it valid to not be initialized off the bat.

### Actual Result

The line 
```
this.#rand ??= Math.random();
```

Results in error
> Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined. 5:5 - 5:15

### Additional Info

Replacing assignment with
```
this.#rand = this.#rand ?? Math.random();
```

No longer produces the error, despite being logically the same, although this will fail on unrelated [logical-assignment-operators](https://eslint.org/docs/latest/rules/logical-assignment-operators).

I suspect the key issues here are combination of a private field, and `exactOptionalPropertyTypes=true` in tsconfig.

**Comment by JacobLey** [comment]
Possibly related previous issue: https://github.com/typescript-eslint/typescript-eslint/issues/4513

**Comment by kirkwaiblinger** [comment]
Just to call out an important point in the report, private properties are the difference between reporting and not reporting here; that is to say, the following (correctly) does not report
```ts
class ConsistentRand {
  rand?: number;

  getCachedRand() {
    this.rand ??= Math.random();
    return this.rand;
  }
}
```

**Comment by kirkwaiblinger** [comment]
Also, see #6635 for the issue which addressed `??=`/`exactOptionalPropertyTypes` previously.

**Comment by JacobLey** [comment]
I'm far from an ESTree expert, but the fix for that issue included a line
```
if (property.type === AST_NODE_TYPES.Identifier) {
```

When switching to private field, the ESTree appears to mark it as a `PrivateIdentifier` (rather than an `Identifier`)

I wonder if the fix is as trivial as expanding that check to include both types...

Doing some hacky debugging locally, that seems to help, but then the following
```
const propertyType = objectType.getProperty(property.name);
```
Results in undefined...

Because the "name" is `rand` but the field is `#rand` (and not addressable via classic string reference). I suspect there is _some_ equivelent on that interface for `getPrivateProperty` or something, but it is not obvious to me...

Maybe `services.getTypeAtLocation(property)` (similar to above) is the trick?
Grasping at straws now...

**Comment by kirkwaiblinger** [comment]
I have no prior knowledge of this specific code area, but I would say it certainly sounds like you are on the right track! 🙂

**Comment by kirkwaiblinger** [comment]
cc @yeonjuan

**Comment by yeonjuan** [comment]
Hi @JacobLey 
Maybe, You could use `checker.getPrivateIdentifierPropertyOfType` to get the type of the private member.

```ts
const tsNode = services.esTreeNodeToTSNodeMap.get(property);
const name = tsNode.getText();

 const privatePropertyType = checker.getPrivateIdentifierPropertyOfType(
        objectType,
        name,
        tsNode,
      );
```
